pkg/handler/instrumented_enqueue_object.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2020 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package handler
+
+import (
+	"github.com/operator-framework/operator-sdk/pkg/handler/internal/metrics"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/util/workqueue"
+
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var enqueueLog = logf.Log.WithName("enqueue_object_event_handler")
+
+// InstrumentedEnqueueRequestForObject wraps controller-runtime handler for "EnqueueRequestForObject", and
+// sets up primary resource metrics on event handlers.
+type InstrumentedEnqueueRequestForObject struct {
+	handler.EnqueueRequestForObject
+}
+
+// Create implements EventHandler, and creates the metrics.
+func (h InstrumentedEnqueueRequestForObject) Create(e event.CreateEvent, q workqueue.RateLimitingInterface) {
+	if e.Meta == nil {
+		enqueueLog.Error(nil, "CreateEvent received with no metadata", "event", e)
+		return
+	}
+	setResourceMetric(e.Meta, e.Object)
+	h.EnqueueRequestForObject.Create(e, q)
+}
+
+// Update implements EventHandler, and updates the metrics.
+func (h InstrumentedEnqueueRequestForObject) Update(e event.UpdateEvent, q workqueue.RateLimitingInterface) {
+	if e.MetaOld != nil {
+		setResourceMetric(e.MetaOld, e.ObjectOld)
+	} else {
+		enqueueLog.Error(nil, "UpdateEvent received with no old metadata", "event", e)
+	}
+	if e.MetaNew != nil {
+		setResourceMetric(e.MetaNew, e.ObjectNew)
+	} else {
+		enqueueLog.Error(nil, "UpdateEvent received with no new metadata", "event", e)
+	}
+	h.EnqueueRequestForObject.Update(e, q)
+}
+
+// Delete implements EventHandler, and deletes metrics.
+func (h InstrumentedEnqueueRequestForObject) Delete(e event.DeleteEvent, q workqueue.RateLimitingInterface) {
+	if e.Meta == nil {
+		enqueueLog.Error(nil, "DeleteEvent received with no metadata", "event", e)
+		return
+	}
+	deleteResourceMetric(e.Meta, e.Object)
+	h.EnqueueRequestForObject.Delete(e, q)
+}
+
+func setResourceMetric(metadata metav1.Object, obj runtime.Object) {
+	labels := getResourceLabels(metadata, obj)
+	m, _ := metrics.ResourceCreatedAt.GetMetricWith(labels)
+	m.Set(float64(metadata.GetCreationTimestamp().UTC().Unix()))
```

Comment:
Instead of doing all of the nil checking in the Create/Update/Delete functions, how about centralizing those here?
```suggestion
	if metadata != nil && obj != nil {
		labels := getResourceLabels(metadata, obj)
		m, _ := metrics.ResourceCreatedAt.GetMetricWith(labels)
		m.Set(float64(metadata.GetCreationTimestamp().UTC().Unix()))
	}
```